### ISSUE TITLE
Make the Bot save the leaderboard by user ids

### ISSUE THREAD

**Issue by Wissididom** [GeorgeCiesinski/poke-guesser-bot#6]
# Describe the new feature
The Users should be saved as ids because usernames and discriminators could be changed. This could lead to two people scoring for the same person or having to start over after a name change where the old name still remains on the leaderboard.

## Planning
One could get the id by msg.author.id (So i think replacing msg.author.username with msg.author.id for $catch, $position and $leaderboard should be enough).
Changing function ```addScore(authorName)``` to ```function addScore(authorId)``` and each occurance of it inside the function.

### Example

```
// Example 1 for adding the score by id with the changed addScore function below API
leaderBoard.addScore(msg.author.id);
// Example 2 for "If on the first element, and element exists, create champion"
{name: `🏆 ${i+1}. <@!${items[i][0]}>`, value: `${items[i][1]} pokémon caught.`},
// Example 3 for "If on element 1-4, and element exists, create new elite four member"
leaderboardEmbed.addField(`🏅 ${i+1}. <@!${items[i][0]}>`, `${items[i][1]} pokémon caught.`);
// Example 4 (untested) for getting a Username with discriminator by id
message.guild.members.get(userID).user.tag; // Example from the Documentation: hydrabolt#0001
// Example 5
const userIds = Array.from(items, user => user[0]);
const userPosition = userIds.indexOf(msg.author.id) + 1;
```

## New Data Structure

Currently, the Replit Database has a key `leaderboard` which stores the leaderboard in the below data structure: 

```
const leaderboard = {
  "user.name(1)":  score,
  "user.name(2)":  score,
};
```

This needs to be changed to 

```
const leaderboard = {
  "user.id(1)":  score,
  "user.id(2)":  score,
};
```

## Required Testing

- [x] Confirm that the user's name can be extrapolated from the user's id. Proof required.

This step is important for the `leaderboard.showLeaderboard()` function which currently relies on the stored names in the `leaderboard` dictionary. Currently, this dictionary is extracted from Replit Database, and used to create an array called `items`, which contains further arrays with the structure `[username, score]`. This username is inserted into the leaderboard by iterating through each item in the items array, and calling `item[0]`. Following the change, this will only yield the id, and not the username. There should be a discord collection in `guild.members` which should hopefully store the members by id, and should have a name key. The `configure.roleExists()` and `configure.channelExists()` functions search the guild's roles and channels collections, and contain code that might be similar to this step. 

This is required for the `leaderboard.findUsername` function in **Required Changes**. Don't proceed with any changes without testing this first.

## Required Changes

- [x] `leaderboard.addScore()` - needs to be updated to store data using New Data Structure
- [x] `leaderboard.showLeaderboard()` - needs to be updated to call a new function for each `user.id` to get the user's current username. Should also handle any event where the id does not yield a user (banned users, users left server, etc). Those users shouldn't be included in the leaderboard and should be removed with `leaderboard.removeUser()`
- [x] `leaderboard.showLeaderboard()` - also needs to output a dummy leaderboard if debugging is enabled
- [x] CREATE `leaderboard.sanitizeLeaderboard()` - Sanitize a leaderboard by removing any users that don't exist in the guild
- [x] CREATE `leaderboard.findUser()` function which takes in the `user.id` or `msg.author.id` and searches through the discord guild members cache to retrieve the user and return object
- [x] `leaderboard.position()` - needs to be updated to search the userNames array using the `user.id` instead of `msg.author.username`
- [x] `leaderboard.newChampionship()` - needs to be updated to find the winner's username by using the `leaderboard.findUsername()` function. 
- [x] `dummyLeaderboard()` - Delete

**Comment by GeorgeCiesinski** [comment]
@Wissididom I have updated this issue to track all of the required changes. Sorry if this seems a bit over complicated. I went through each of the functions that use the leaderboard dictionary and anticipated errors that would come up from changes. Before either of us work on this, would you be able to take a look during your free time to see if I missed anything, or if you have a better idea of how this change can be made?

**Comment by Wissididom** [comment]
I just found something that is missing in the listing above:
index.js:208: Change `leaderBoard.addScore(msg.author.username);` should be replaced with `leaderBoard.addScore(msg.author.id);`
I think there is really no applicable other choice than updating all of these functions. We could give the functions in leaderboard.js a user object and then having that function extract the id but that wouldn't make any difference to only giving the functions the id because at every other function than addScore and maybe position we would also need to use the findUsername function.

**Comment by GeorgeCiesinski** [comment]
For point: 

> Confirm that the user's name can be extrapolated from the user's id. Proof required.

@Wissididom 's `findUser` has been tested and correctly returns the username from the id.

**Comment by GeorgeCiesinski** [comment]
For point: 

> leaderboard.showLeaderboard - Should also handle any event where the id does not yield a user (banned users, users left server, etc).

Created `leaderboard.sanitizeLeaderboard` which iterates through the leaderboard and removes any non-existent users. Called at the start of showLeaderboard.

**Comment by GeorgeCiesinski** [comment]
Adding useful link for async issue experienced when trying to sanitizeLeaderboard: 

https://stackoverflow.com/questions/14220321/how-to-return-the-response-from-an-asynchronous-call

**Comment by GeorgeCiesinski** [comment]
# addScore

- Modified to store `id: score`

## Proof

<img width="506" alt="Screen Shot 2021-07-30 at 8 17 58 PM" src="https://user-images.githubusercontent.com/7014211/127722715-c4759286-d1d3-4e88-85af-209ae137e025.png">

<img width="408" alt="Screen Shot 2021-07-30 at 8 18 20 PM" src="https://user-images.githubusercontent.com/7014211/127722717-f352d8a0-300f-4734-8443-9f8dd6b2efdd.png">


# showLeaderboard

- Modified to work with ids instead of names
- Utilizes `sanitizeLeaderboard` function to remove non-existent users
- Added `!leaderboard debug` command that runs `showLeaderboard` in debug mode. This mode calls `generateLeaderboard` to generate 20 additional users (named user1, user2...) and adds them to the leaderboard. This makes it possible to test leaderboard embed changes.
- At the end of the function, the sanitizeLeaderboard is stored to replace the existing leaderboard, permanently removing non-existent users

## Proof - Empty Leaderboard

<img width="420" alt="Screen Shot 2021-07-30 at 8 16 21 PM" src="https://user-images.githubusercontent.com/7014211/127722635-d7f4e424-7597-4b51-9255-ae23d6f907bb.png">

## Proof - Leaderboard Debug Mode

- Appends fake users to leaderboard without deleting existing users. Not recommended on live bot.

<img width="236" alt="Screen Shot 2021-07-30 at 8 21 58 PM" src="https://user-images.githubusercontent.com/7014211/127722807-449fff72-d499-41ac-b969-98c14f689729.png">

<img width="454" alt="Screen Shot 2021-07-30 at 8 22 26 PM" src="https://user-images.githubusercontent.com/7014211/127722818-fd677a00-0cac-4a8e-aefb-de20eeab4c52.png">

## Proof - sanitizeLeaderboard

- Running `!leaderboard` after `!leaderboard debug` results in the non-existent added users being removed so the leaderboard generates correctly with real users only

<img width="404" alt="Screen Shot 2021-07-30 at 8 24 14 PM" src="https://user-images.githubusercontent.com/7014211/127722849-1924a480-85e9-476c-9f58-c0bcf1563e96.png">

**Comment by GeorgeCiesinski** [comment]
Resolved in Pull Request #16
